staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/resourceclaim.go

```diff
@@ -78,6 +80,70 @@ func Name(pod *v1.Pod, podClaim *v1.PodResourceClaim) (name *string, mustCheckOw
 	}
 }
 
+// NewNameLookup returns an object which handles determining the name of
+// a ResourceClaim. In contrast to the stand-alone Name it is compatible
+// also with Kubernetes < 1.28.
+//
+// Providing a client is optional. If none is available, then code can pass nil
+// and users can set the DRA_WITH_DETERMINISTIC_RESOURCE_CLAIM_NAMES env
+// variable to an arbitrary non-empty value to use the naming from Kubernetes <
+// 1.28.
+func NewNameLookup(client kubernetes.Interface) *Lookup {
+	return &Lookup{client: client}
+}
+
+// Lookup stores the state which is necessary to look up ResourceClaim names.
+type Lookup struct {
+	client       kubernetes.Interface
+	usePodStatus *bool
+}
+
+// Name is a variant of the stand-alone Name with support also for Kubernetes < 1.28.
+func (l *Lookup) Name(pod *v1.Pod, podClaim *v1.PodResourceClaim) (name *string, mustCheckOwner bool, err error) {
+	if l.usePodStatus == nil {
+		if value, _ := os.LookupEnv("DRA_WITH_DETERMINISTIC_RESOURCE_CLAIM_NAMES"); value != "" {
+			l.usePodStatus = ptr.To(false)
+		} else if l.client != nil {
+			// Check once. This does not detect upgrades or
+			// downgrades, but that is good enough for the simple
+			// test scenarios that the Kubernetes < 1.28 support is
+			// meant for.
+			info, err := l.client.Discovery().ServerVersion()
+			if err != nil {
+				return nil, false, fmt.Errorf("look up server version: %v", err)
+			}
+			if info.Major == "" {
+				// Fake client...
+				l.usePodStatus = ptr.To(true)
+			} else {
+				major, err := strconv.Atoi(info.Major)
+				if err != nil {
+					return nil, false, fmt.Errorf("parsing major version: %v", err)
+				}
+				minor, err := strconv.Atoi(info.Minor)
+				if err != nil {
+					return nil, false, fmt.Errorf("parsing minor version: %v", err)
+				}
+				l.usePodStatus = ptr.To(major > 1 || minor >= 28)
+			}
+		}
+	}
+
+	if *l.usePodStatus {
+		return Name(pod, podClaim)
```

Comment:
Ah ... ok, I see that the  relationship is the inverse of what I was expecting. As a matter of interest, do we expect that `Name` will be the function used moving forward, or would we rather have clients depend on the `Lookup` type (or possibly an interface)?

Reply:
I see `NewNameLookup` as an interim solution and want to remove it again once 1.27 is no longer supported, upstream or by this package (whatever happens sooner).

I was tempted to clarify that by adding a `Deprecated: use Name instead of you don't need to support Kubernetes 1.27`, but it felt odd to add something that is deprecated right out of the gate.